File: TradeReplayAssistant/Domain/TradeModels.swift
import Foundation

enum TradeDirection: String, Codable, CaseIterable, Identifiable, Sendable {
    case long = "做多"
    case short = "做空"

    var id: String { rawValue }
}

enum SyncState: Equatable, Sendable {
    case idle
    case syncing
    case synced(Date)
    case failed(String)
    case conflict(Int)
}

enum StatsRangeWindow: String, CaseIterable, Identifiable, Sendable {
    case last7Days = "近7天"
    case last30Days = "近30天"
    case yearToDate = "年内"

    var id: String { rawValue }

    func interval(referenceDate: Date = .now, calendar: Calendar = .current) -> DateInterval {
        switch self {
        case .last7Days:
            let start = calendar.date(byAdding: .day, value: -6, to: referenceDate) ?? referenceDate
            return DateInterval(start: calendar.startOfDay(for: start), end: referenceDate)
        case .last30Days:
            let start = calendar.date(byAdding: .day, value: -29, to: referenceDate) ?? referenceDate
            return DateInterval(start: calendar.startOfDay(for: start), end: referenceDate)
        case .yearToDate:
            let components = calendar.dateComponents([.year], from: referenceDate)
            let start = calendar.date(from: components) ?? referenceDate
            return DateInterval(start: start, end: referenceDate)
        }
    }
}

struct TradeSnapshot: Identifiable, Codable, Hashable, Sendable {
    var id: UUID
    var symbol: String
    var direction: TradeDirection
    var openTime: Date
    var closeTime: Date
    var openPrice: Double
    var closePrice: Double
    var positionSize: Double
    var fee: Double
    var pnl: Double
    var strategyTag: String
    var mistakeTag: String
    var notes: String
    var createdAt: Date
    var updatedAt: Date
    var deletedAt: Date?

    var isWin: Bool { pnl > 0 }

    var holdingMinutes: Int {
        max(Int(closeTime.timeIntervalSince(openTime) / 60), 0)
    }

    var isDeleted: Bool {
        guard let deletedAt else { return false }
        return deletedAt <= .now
    }
}

struct TradeDraft: Sendable {
    var symbol: String = ""
    var direction: TradeDirection = .long
    var openTime: Date = .now
    var closeTime: Date = .now
    var openPrice: Double = 0
    var closePrice: Double = 0
    var positionSize: Double = 1
    var fee: Double = 0
    var strategyTag: String = "趋势"
    var mistakeTag: String = ""
    var notes: String = ""

    func toSnapshot(id: UUID, createdAt: Date, updatedAt: Date) -> TradeSnapshot {
        let directionMultiplier: Double = direction == .long ? 1 : -1
        let gross = (closePrice - openPrice) * positionSize * directionMultiplier
        let pnl = gross - fee

        return TradeSnapshot(
            id: id,
            symbol: symbol.trimmingCharacters(in: .whitespacesAndNewlines).uppercased(),
            direction: direction,
            openTime: openTime,
            closeTime: closeTime,
            openPrice: openPrice,
            closePrice: closePrice,
            positionSize: positionSize,
            fee: fee,
            pnl: pnl,
            strategyTag: strategyTag.trimmingCharacters(in: .whitespacesAndNewlines),
            mistakeTag: mistakeTag.trimmingCharacters(in: .whitespacesAndNewlines),
            notes: notes.trimmingCharacters(in: .whitespacesAndNewlines),
            createdAt: createdAt,
            updatedAt: updatedAt,
            deletedAt: nil
        )
    }
}

struct TradeSummary: Sendable {
    var totalTrades: Int
    var winRate: Double
    var totalPnL: Double
    var averageWin: Double
    var averageLoss: Double
    var profitFactor: Double
    var maxDrawdown: Double

    static let empty = TradeSummary(
        totalTrades: 0,
        winRate: 0,
        totalPnL: 0,
        averageWin: 0,
        averageLoss: 0,
        profitFactor: 0,
        maxDrawdown: 0
    )
}

struct WeeklyReport: Sendable {
    var weekInterval: DateInterval
    var summary: TradeSummary
    var goodPractices: [String]
    var improvementAreas: [String]
    var nextWeekActions: [String]

    static func empty(for interval: DateInterval) -> WeeklyReport {
        WeeklyReport(
            weekInterval: interval,
            summary: .empty,
            goodPractices: ["本周暂无交易数据"],
            improvementAreas: ["先完成至少 5 笔可复盘交易"],
            nextWeekActions: ["交易结束后 10 分钟内填写复盘记录"]
        )
    }
}

extension TradeSnapshot {
    static func demoSeed(referenceDate: Date = .now) -> [TradeSnapshot] {
        let calendar = Calendar.current

        func makeDayOffset(_ day: Int, hour: Int, minute: Int) -> Date {
            let dayDate = calendar.date(byAdding: .day, value: day, to: referenceDate) ?? referenceDate
            return calendar.date(
                bySettingHour: hour,
                minute: minute,
                second: 0,
                of: dayDate
            ) ?? dayDate
        }

        return [
            TradeSnapshot(
                id: UUID(),
                symbol: "AAPL",
                direction: .long,
                openTime: makeDayOffset(-5, hour: 10, minute: 2),
                closeTime: makeDayOffset(-5, hour: 11, minute: 21),
                openPrice: 188.2,
                closePrice: 190.1,
                positionSize: 100,
                fee: 8,
                pnl: 182,
                strategyTag: "突破",
                mistakeTag: "",
                notes: "量价齐升后跟随，止盈执行良好。",
                createdAt: makeDayOffset(-5, hour: 11, minute: 22),
                updatedAt: makeDayOffset(-5, hour: 11, minute: 25),
                deletedAt: nil
            ),
            TradeSnapshot(
                id: UUID(),
                symbol: "TSLA",
                direction: .short,
                openTime: makeDayOffset(-3, hour: 9, minute: 44),
                closeTime: makeDayOffset(-3, hour: 10, minute: 15),
                openPrice: 171.5,
                closePrice: 173.0,
                positionSize: 80,
                fee: 7,
                pnl: -127,
                strategyTag: "回落",
                mistakeTag: "止损迟疑",
                notes: "触发止损后仍犹豫，扩大亏损。",
                createdAt: makeDayOffset(-3, hour: 10, minute: 16),
                updatedAt: makeDayOffset(-3, hour: 10, minute: 20),
                deletedAt: nil
            ),
            TradeSnapshot(
                id: UUID(),
                symbol: "NVDA",
                direction: .long,
                openTime: makeDayOffset(-1, hour: 13, minute: 2),
                closeTime: makeDayOffset(-1, hour: 14, minute: 10),
                openPrice: 943.0,
                closePrice: 951.8,
                positionSize: 20,
                fee: 6,
                pnl: 170,
                strategyTag: "趋势",
                mistakeTag: "",
                notes: "顺势加仓后分批止盈，执行稳定。",
                createdAt: makeDayOffset(-1, hour: 14, minute: 11),
                updatedAt: makeDayOffset(-1, hour: 14, minute: 14),
                deletedAt: nil
            )
        ]
    }
}

File: TradeReplayAssistantAndroid/app/src/main/java/com/punglegod/trade/domain/TradeModels.kt
package com.punglegod.trade.domain

import kotlinx.serialization.Serializable
import kotlin.math.abs
import kotlin.math.max

@Serializable
enum class TradeDirection {
    LONG,
    SHORT;

    val label: String
        get() = when (this) {
            LONG -> "做多"
            SHORT -> "做空"
        }
}

@Serializable
data class TradeSnapshot(
    val id: String,
    val symbol: String,
    val direction: TradeDirection,
    val openTimeMillis: Long,
    val closeTimeMillis: Long,
    val openPrice: Double,
    val closePrice: Double,
    val positionSize: Double,
    val fee: Double,
    val pnl: Double,
    val strategyTag: String,
    val mistakeTag: String,
    val notes: String,
    val createdAtMillis: Long,
    val updatedAtMillis: Long,
    val deletedAtMillis: Long? = null
) {
    val isDeleted: Boolean get() = deletedAtMillis != null
    val holdingMinutes: Int get() = max(((closeTimeMillis - openTimeMillis) / 60_000).toInt(), 0)
}

@Serializable
data class TradeDraft(
    val symbol: String = "",
    val direction: TradeDirection = TradeDirection.LONG,
    val openTimeMillis: Long = System.currentTimeMillis(),
    val closeTimeMillis: Long = System.currentTimeMillis(),
    val openPrice: Double = 0.0,
    val closePrice: Double = 0.0,
    val positionSize: Double = 1.0,
    val fee: Double = 0.0,
    val strategyTag: String = "趋势",
    val mistakeTag: String = "",
    val notes: String = ""
) {
    fun toSnapshot(id: String, createdAtMillis: Long, updatedAtMillis: Long): TradeSnapshot {
        val directionMultiplier = if (direction == TradeDirection.LONG) 1.0 else -1.0
        val gross = (closePrice - openPrice) * positionSize * directionMultiplier
        val normalizedSymbol = symbol.trim().uppercase()

        return TradeSnapshot(
            id = id,
            symbol = normalizedSymbol,
            direction = direction,
            openTimeMillis = openTimeMillis,
            closeTimeMillis = closeTimeMillis,
            openPrice = openPrice,
            closePrice = closePrice,
            positionSize = positionSize,
            fee = fee,
            pnl = gross - fee,
            strategyTag = strategyTag.trim(),
            mistakeTag = mistakeTag.trim(),
            notes = notes.trim(),
            createdAtMillis = createdAtMillis,
            updatedAtMillis = updatedAtMillis,
            deletedAtMillis = null
        )
    }
}

enum class StatsRangeWindow(val label: String) {
    LAST_7_DAYS("近7天"),
    LAST_30_DAYS("近30天"),
    YEAR_TO_DATE("年内");

    fun startMillis(referenceMillis: Long): Long {
        val dayMillis = 24L * 60 * 60 * 1_000
        return when (this) {
            LAST_7_DAYS -> referenceMillis - (6 * dayMillis)
            LAST_30_DAYS -> referenceMillis - (29 * dayMillis)
            YEAR_TO_DATE -> {
                val cal = java.util.Calendar.getInstance().apply { timeInMillis = referenceMillis }
                cal.set(java.util.Calendar.MONTH, java.util.Calendar.JANUARY)
                cal.set(java.util.Calendar.DAY_OF_MONTH, 1)
                cal.set(java.util.Calendar.HOUR_OF_DAY, 0)
                cal.set(java.util.Calendar.MINUTE, 0)
                cal.set(java.util.Calendar.SECOND, 0)
                cal.set(java.util.Calendar.MILLISECOND, 0)
                cal.timeInMillis
            }
        }
    }
}

data class TradeSummary(
    val totalTrades: Int,
    val winRate: Double,
    val totalPnL: Double,
    val averageWin: Double,
    val averageLoss: Double,
    val profitFactor: Double,
    val maxDrawdown: Double
) {
    companion object {
        val Empty = TradeSummary(
            totalTrades = 0,
            winRate = 0.0,
            totalPnL = 0.0,
            averageWin = 0.0,
            averageLoss = 0.0,
            profitFactor = 0.0,
            maxDrawdown = 0.0
        )
    }
}

data class WeeklyReport(
    val weekStartMillis: Long,
    val weekEndMillis: Long,
    val summary: TradeSummary,
    val goodPractices: List<String>,
    val improvementAreas: List<String>,
    val nextWeekActions: List<String>
) {
    companion object {
        fun empty(referenceMillis: Long): WeeklyReport {
            val dayMillis = 24L * 60 * 60 * 1_000
            return WeeklyReport(
                weekStartMillis = referenceMillis - (6 * dayMillis),
                weekEndMillis = referenceMillis,
                summary = TradeSummary.Empty,
                goodPractices = listOf("本周暂无交易数据"),
                improvementAreas = listOf("先完成至少 5 笔可复盘交易"),
                nextWeekActions = listOf("交易结束后 10 分钟内填写复盘记录")
            )
        }
    }
}

fun List<String>.toLabeledTop3(fallback: List<String>): List<String> {
    val clean = map { it.trim() }.filter { it.isNotEmpty() }
    if (clean.isEmpty()) return fallback
    return clean.groupingBy { it }
        .eachCount()
        .entries
        .sortedWith(compareByDescending<Map.Entry<String, Int>> { it.value }.thenBy { it.key })
        .take(3)
        .map { it.key }
}

fun Double.formatSignedPnL(): String {
    val rounded = "%.2f".format(abs(this))
    return if (this >= 0) "+$rounded" else "-$rounded"
}
